keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+ResourceClassParameters type. Instead of individual requests, one additional
+selector can be specified there which then also must be true for all individual
+requests made with that class:
+
+```yaml
+kind: ResourceClassParameters
+apiVersion: resource.k8s.io/v1alpha2
+
+metadata:
+  name: someArbitraryName
+
+generatedFrom:
+  name: gpu-parameters
+  kind: CardClassParameters
+  apiGroup: dra.example.com
+  uid: foobar-uid
+
+vendorParameters:
+  ...
+
+filters:
+- driverName: cards.dra.example.com
+  partitionable:
+    selector: |-
+      quantities["memory"] <= "16Gi"
+```
+
+In this example, the additional selector expression limits users of this class
+to smaller sub-partitions. Together with limiting the number of claims that
+users are allowed to create for this class (see resource quotas in the core
+KEP) this can ensure that users do not consume too much resources. Allowing
+resource quotas that are based on resource attributes may be a useful future
+enhancement.
+
+### Communicating allocation to the DRA driver
+
+The scheduler decides which resources to use for a claim and how much of
+them. It also needs to pass through the opaque vendor parameters, if there are
+any. All of this information gets stored in the allocation result inside the
+ResourceClaim status. For the example above, the result produced by the
+scheduler is simply the list of IDs of the selected sub-partitions:
+
+```yaml
+# Matches with the SemanticResourceHandle Go type defined below.
+vendorClassParameters:
+  ...
+vendorClaimParameters:
+  kind: CardParameters
+  apiVersion: dra.example.com/v1alpha1
+  metadata:
+    name: my-parameters
+    namespace: user-namespace
+    uid: foobar-uid
+  ...
+  spec:
+    count: 2
+    minimumRuntimeVersion: v12.0.0
+    minimumMemory: 8Gi
+    sharing:
+      strategy: TimeSliced
+
+nodeName: worker-1
+partitionable:
+  instances:
+  - partitionable-gpu-2/half-and-quarters/second-half/quarter-0
+  - partitionable-gpu-2/half-and-quarters/second-half/quarter-1
+```
+
+## Design Details
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables semantic parameters for claims
+using this class:
+
+```go
+type ResourceClass struct {
+    ...
+
+    // If (and only if) allocation of claims using this class is handled
+    // via semantic parameters, then SemanticParameters must be set to true.
+    SemanticParameters bool
+}
+```
+
+### NodeResourceSlice
+
+For each node, one or more NodeResourceSlice objects get created. The kubelet
+publishes them with the node as the owner, so they get deleted when a node goes
+down and then gets removed.
+
+All list types are atomic because that makes tracking the owner for
+server-side-apply (SSA) simpler. Patching individual list elements is not
+needed and there is a single owner (kubelet).
+
+```go
+// NodeResourceSlice provides information about available
+// resources on individual nodes.
+type NodeResourceSlice struct {
```

Comment:
I'd like to revisit the name `NodeResourceSlice` during implementation. This still doesn't feel like the right term to capture what we are advertising here.

I know that `Slice` here is meant to denote that there is a `List` of `NodeResources`, but I can't seem to get past the connotation that this struct instead represents a "slice" (as in a partition or piece) of a single resource on a node, i.e. "a node resource slice".